src/Fantomas/CodeFormatterImpl.fs

```diff
@@ -359,29 +354,26 @@ let isValidFSharpCode (checker: FSharpChecker) (parsingOptions: FSharpParsingOpt
             let! ast = parse checker parsingOptions formatContext
 
             let isValid =
-                ast |> Array.map fst |> Array.forall isValidAST
+                ast
+                |> Array.map (fun (a, _, _) -> a)
+                |> Array.forall isValidAST
 
             return isValid
         with _ -> return false
     }
 
-let formatWith ast defines formatContext config =
+let formatWith ast defines hashTokens formatContext config =
     let moduleName =
         Path.GetFileNameWithoutExtension formatContext.FileName
 
-    let input =
+    let sourceCode, hasInput =
         if String.IsNullOrWhiteSpace formatContext.Source
-        then None
-        else Some formatContext.Source
-
-    // Use '\n' as the new line delimiter consistently
-    // It would be easier for F# parser
-    let sourceCode = defaultArg input String.Empty
-    let normalizedSourceCode = String.normalizeNewLine sourceCode
+        then String.Empty, false
+        else formatContext.Source, true
```

Comment:
This feels like a slightly broken domain. `sourceCode, hasInput` is really a `string option` that you're representing as a `string * bool`. This makes the `IsNullOrWhitespace` check on line 390 a bit harder to understand.